hw/xbox/mcpx_apu.c

```diff
@@ -403,25 +408,42 @@ static const MemoryRegionOps vp_ops = {
     .write = vp_write,
 };
 
-static void scratch_rw(hwaddr sge_base, unsigned int max_sge,
+static void scratch_rw(MCPXAPUState *d,
+                       hwaddr sge_base, unsigned int max_sge,
                        uint8_t *ptr, uint32_t addr, size_t len, bool dir)
 {
-    int i;
-    for (i = 0; i < len; i++) {
-        unsigned int entry = (addr + i) / TARGET_PAGE_SIZE;
-        assert(entry < max_sge);
+    unsigned int page_entry = addr / TARGET_PAGE_SIZE;
+    unsigned int offset_in_page = addr % TARGET_PAGE_SIZE;
+    unsigned int bytes_to_copy = TARGET_PAGE_SIZE - offset_in_page;
+
+    while (len > 0) {
+        assert(page_entry < max_sge);
+
         uint32_t prd_address = ldl_le_phys(&address_space_memory,
-                                           sge_base + entry * 8 + 0);
+                                           sge_base + page_entry * 8 + 0);
         /* uint32_t prd_control = ldl_le_phys(&address_space_memory,
-                                            sge_base + entry * 8 + 4); */
+                                            sge_base + page_entry * 8 + 4); */
 
-        hwaddr paddr = prd_address + (addr + i) % TARGET_PAGE_SIZE;
+        hwaddr paddr = prd_address + offset_in_page;
```

Comment:
Hopefully paddr is in [0, 128M)... Might be good to have a sanity check here before we access RAM?

Reply:
Yeah, I'll add this. Good catch.